### ISSUE TITLE
[sklearn#34436] Birch: subcluster radius is numerically unstable far from the origin — clustering is not translation-invariant

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#325]
**Mirrored from [scikit-learn#34436](https://github.com/scikit-learn/scikit-learn/issues/34436)**

Original title: Birch: subcluster radius is numerically unstable far from the origin — clustering is not translation-invariant
Original labels: Bug, Needs Triage
Created: 2026-07-06T10:12:27Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug and give evidence about its user-facing impact

**``Birch``**  derives each subcluster's squared radius from the **non-centered** second moment, `r² = Σᵢ‖xᵢ‖²/n − ‖c‖²` (`_CFSubcluster.merge_subcluster` in `sklearn/cluster/_birch.py`). Both terms scale like ‖offset‖², so far from the origin they are large and nearly equal, and their float64 difference loses the unit-scale variance to **catastrophic cancellation**. `r²` then reads noise (often negative → clamped to 0), so `threshold` no longer bounds the leaf radius and the CF-tree silently changes structure. Concretely, a **pure translation** of the input — which cannot change Birch's result in exact arithmetic — changes the number of subclusters.

### Steps/Code to Reproduce

```python
import numpy as np, sklearn
from sklearn.cluster import Birch

rng = np.random.RandomState(0)
X = np.vstack([rng.randn(200, 2), rng.randn(200, 2) + [15.0, 0.0]])  # two tight blobs, spread ~1

base = len(Birch(threshold=1.0, n_clusters=None).fit(X).subcluster_centers_)
print(f"sklearn {sklearn.__version__}  baseline (offset 0): {base} subclusters")
for off in (1e5, 1e6, 1e7, 1e8):
    n = len(Birch(threshold=1.0, n_clusters=None).fit(X + off).subcluster_centers_)
    print(f"  offset={off:.0e}: {n:4d}  {'OK' if n == base else 'DEGRADED'}")
```

### Expected Results

The subcluster count is invariant under translation (Birch's `threshold` test is translation-equivariant in exact arithmetic): `12` at every offset.

### Actual Results

```
sklearn 1.9.0  baseline (offset 0): 12 subclusters
  offset=1e+05:   12  OK
  offset=1e+06:   13  DEGRADED
  offset=1e+07:   11  DEGRADED
  offset=1e+08:   99  DEGRADED
```

### Versions

```shell
System:
    python: 3.12.13 (main, Mar 24 2026, 22:49:22) [Clang 22.1.1 ]
   machine: Linux-7.0.14-201.fc44.x86_64-x86_64-with-glibc2.43

Python dependencies:
      sklearn: 1.9.0
        numpy: 2.5.1
        scipy: 1.18.0
       Cython: None
       joblib: 1.5.3
threadpoolctl: 3.6.0

Built with OpenMP: True

threadpoolctl info:
       user_api: blas
   internal_api: openblas
    num_threads: 16
         prefix: libscipy_openblas
        version: 0.3.33.112.0
threading_layer: pthreads
   architecture: Haswell

       user_api: openmp
   internal_api: openmp
    num_threads: 16
         prefix: libgomp
```

### Interest in fixing the bug

Yes, I'd be glad to work on the PR. The cancellation is in `new_ss / new_n − new_sq_norm`, from accumulating `squared_sum_` as raw `Σ‖x‖²`. The standard remedy: store the

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/98 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
